File: backend/app/schemas/product.py
from pydantic import BaseModel, ConfigDict, Field

from app.schemas.category import CategoryRead


class ProductBase(BaseModel):
    name: str = Field(min_length=1, max_length=255)
    description: str = Field(min_length=1)
    category_id: int
    brand: str | None = Field(default=None, max_length=100)
    price: int = Field(ge=0)
    rating: float | None = Field(default=None, ge=0, le=5)
    stock: int | None = Field(default=None, ge=0)
    weight_g: int | None = Field(default=None, ge=0)
    battery_hours: int | None = Field(default=None, ge=0)
    screen_size: float | None = Field(default=None, ge=0)
    memory_gb: int | None = Field(default=None, ge=0)
    storage_gb: int | None = Field(default=None, ge=0)
    tags: list[str] = Field(default_factory=list)
    thumbnail_url: str | None = Field(default=None, max_length=500)


class ProductCreate(ProductBase):
    pass


class ProductUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=255)
    description: str | None = Field(default=None, min_length=1)
    category_id: int | None = None
    brand: str | None = Field(default=None, max_length=100)
    price: int | None = Field(default=None, ge=0)
    rating: float | None = Field(default=None, ge=0, le=5)
    stock: int | None = Field(default=None, ge=0)
    weight_g: int | None = Field(default=None, ge=0)
    battery_hours: int | None = Field(default=None, ge=0)
    screen_size: float | None = Field(default=None, ge=0)
    memory_gb: int | None = Field(default=None, ge=0)
    storage_gb: int | None = Field(default=None, ge=0)
    tags: list[str] | None = None
    thumbnail_url: str | None = Field(default=None, max_length=500)


class ProductRead(BaseModel):
    id: int
    name: str
    description: str
    category: CategoryRead
    brand: str | None
    price: int
    rating: float | None
    stock: int | None
    weight_g: int | None
    battery_hours: int | None
    screen_size: float | None
    memory_gb: int | None
    storage_gb: int | None
    tags: list[str]
    thumbnail_url: str | None = None

    model_config = ConfigDict(from_attributes=True)


class ProductListResponse(BaseModel):
    items: list[ProductRead]
    page: int
    limit: int
    total: int
    has_next: bool

File: frontend/types/product.ts
export type Category = {
  id: number;
  name: string;
  slug: string;
};

export type Product = {
  id: number;
  name: string;
  description: string;
  category: Category;
  brand: string | null;
  price: number;
  rating: number | null;
  stock: number | null;
  weight_g: number | null;
  battery_hours: number | null;
  screen_size: number | null;
  memory_gb: number | null;
  storage_gb: number | null;
  tags: string[];
  thumbnail_url: string | null;
};

export type ProductSort = "price_asc" | "price_desc" | "rating_desc" | "newest";

export type ProductQuery = {
  keyword?: string;
  category_id?: number;
  min_price?: number;
  max_price?: number;
  sort?: ProductSort;
  page?: number;
  limit?: number;
};

export type ProductListResponse = {
  items: Product[];
  page: number;
  limit: number;
  total: number;
  has_next: boolean;
};

export type ProductPayload = {
  name: string;
  description: string;
  category_id: number;
  brand: string | null;
  price: number;
  rating: number | null;
  stock: number | null;
  weight_g: number | null;
  battery_hours: number | null;
  screen_size: number | null;
  memory_gb: number | null;
  storage_gb: number | null;
  tags: string[];
  thumbnail_url: string | null;
};
